Improved Linux detection fixes

XenHat opened the pull request:
This is a re-implementation of #92 with additional fixes for more games, and should be compatible with #130 as I have validated the code against the current upstream database.

The following games have been verified to be correctly detected on **Linux**:

- [x] Factorio (Proton)
- [x] Final Fantasy XIV Online
- [x] Genshin Impact
- [x] Grand Theft Auto: Vice City
- [x] Heaven Burns Red
- [x] Helldivers 2
- [x] Hollow Knight (Native)
- [x] Honkai: Star Rail
- [x] Horizon Zero Dawn
- [x] Journey
- [x] Mafia: The Old Country
- [x] Minecraft
- [x] Satisfactory
- [x] Snowrunner
- [x] STAR WARS™ Battlefront™ II
- [x] The Elder Scrolls IV: Oblivion Remastered
- [x] Wuthering Waves
- [x] X4: Foundations (Proton)

KNOWN ISSUES:

- Red Dead Redemption 2 (Broken)
- Zenless Zone Zero (Needs re-test)
- Untested on Windows, but should work.
- Many Linux Native titles are not detected (Missing from the Discord DB)

Main changes:
 - several performance improvements
 - match against more (malformed?) game database entries

Additional credit:
- Marcel Witte (<wittemar@gmail.com>) for the Java workaround
- Konovalov Nikolay for the more efficient executable name matching

Justsnoopy30 committed:
process: improve linux game detection

Some games are listed in the discord game database with their parent directories
but their launched processes on linux don't always have this included in cmdline.
This fixes that by checking processes with their working directories as well.
For example, this will now detect DOOM Eternal running through proton.

Justsnoopy30 committed:
process: small fix for false positives

XenHat commented on src/process/index.js:
```diff
@@ -47,12 +51,39 @@ export default class ProcessServer {
         if (executables?.some(x => {
           if (x.is_launcher) return false;
-          if (x.name[0] === '>' ? x.name.substring(1) !== toCompare[0] : !toCompare.some(y => x.name === y)) return false;
+          // I can't figure how to not make 'sbin/init' NOT match 'Divinity Original Sin' with a substring match, so this ugly hack have to do for now
```
I'm not really happy with this code, but it works. I'm new to javascript/nodejs and I couldn't get a more elegant solution to work as I cannot wrap my brain around the arrow operator

XenHat commented on src/process/index.js:
```diff
@@ -47,12 +51,39 @@ export default class ProcessServer {
+          // I can't figure how to not make 'sbin/init' NOT match 'Divinity Original Sin' with a substring match, so this ugly hack have to do for now
+          for (const comp of UniqueToCompare) {
+            const asdf = comp.includes('.exe')
+            if (asdf) {
```
These will be renamed

XenHat reviewed (commented):
Added some notes about code I forgot to finalize

XenHat commented:
I've been using this for a few days, seems to be working as intended.
Testing needs to be done on Windows by users of this operating system.

XenHat commented:
Known issue: `The Elder Scrolls IV: Oblivion Remastered` isn't detected properly in the new code. An up-to-date detectable.json works with the code prior to this PR but does not with my new code.

ikaikahub commented:
fixes my issue with running thcrap patched touhou 7, cool pr

XenHat commented on src/process/index.js:
```diff
@@ -32,30 +64,94 @@ export default class ProcessServer {
+            if (known_exe.is_launcher) return false;
+            if (known_exe.name[0] === '>') {
+              if (known_exe.name.substring(1) === toCompare[0]) {
+                return true;
```
Accepted, minecraft is still detected (through Prism Launcher at least.

XenHat commented:
As of today, I cannot reconnect the standalone arrpc to vesktop anymore after clearing the settings, and thus will be unable to keep working on this for the time being.
I will refrain from pointing fingers, but suffice to say, I am very displeased with the state of things.

konovalov-nk commented on src/process/index.js:
```diff
@@ -32,30 +64,96 @@ export default class ProcessServer {
+                toCompare.some((running) => {
+                  // explicit match first
+                  if (known_exe.name === running) {
+                    // log(`Match Level 2: "${name}" via ${known_exe.name} <==> ${running}`)
```
These log comments aren't really necessary to understand what's going on but if here's two options I can suggest:

Option 1: make it `console.debug` instead of `.log`, so we can configure logging levels and skip `debug` by default. This gives you an easy toggle for traces if you need them, and not just a comment.

```suggestion
                    // debug(`Match Level 2: "${name}" via ${known_exe.name} <==> ${running}`)
```

or

```suggestion
```

konovalov-nk commented on src/process/index.js:
Can we refactor this file to something like this?

https://jsfiddle.net/jsq76fyc/ <-- much easier to reason about than quadruple nested `if` blocks 🙂

XenHat replied:
Done 😄

XenHat replied:
@konovalov-nk Done! Implemented at 2d30239d3eec4b0a948c128a3f8aa6238910bccb

neptuwunium commented on src/process/index.js:
```diff
@@ -32,30 +65,95 @@ export default class ProcessServer {
       const path = _path.toLowerCase().replaceAll('\\', '/');
+      if (path.startsWith('c:/windows')) continue // system processes (wine)
+      if (_path.includes('webhelper')) continue; // CEF Processes
+      // TODO: add 'dolphin-emu' to database for linux executable
```
Probably should have a less hacky mechanism to patch existing entries or add/remove them. Vesktop adds one for OBS to enable streamer mode by patching the source as well: https://github.com/Vencord/Vesktop/blob/7b5e1ed4da50c02c9ed3dd976066364d1f19c106/patches/arrpc%403.5.0.patch

XenHat replied:
Done :)

konovalov-nk commented on src/process/index.js:
```diff
@@ -32,30 +155,50 @@ export default class ProcessServer {
+      if (_path.endsWith('/bin/dolphin')) continue; // KDE file manager, not Dolphin Emulator
+      if (!_path.includes('\\') && !_path.includes('/')) continue;
+      if (_path.startsWith('C:\\windows')) continue // Windows processes
+      if (_path.includes('webhelper')) continue; // CEF Processes
```
This feels like 

<img width="476" height="528" alt="Image" src="https://github.com/user-attachments/assets/8fad47b5-a3dd-427f-9d5d-6e5140f7a2c2" />

Why are we even doing this?

The problem we're trying to solve here is: "enumerate the universe of non-games".
If we want to ignore specific paths, we can simply use grep/ripgrep/regex?
This should not be solved with control-flow, it should be data problem + logic.

```js
const IGNORE_RULES = {
  linux: [
    /^\/proc\//,
    /^\/usr\/(bin|sbin|lib)\//,
    /^\/(bin|sbin)\//,
    /systemd/,
    /webhelper/,
    /\/bin\/dolphin$/, // KDE, not emulator
  ],
  windows: [
    /^C:\\Windows/i,
    path => path.includes('\\') && !path.includes('/'), // pure win path
  ]
};

// That's literally it:
function shouldIgnoreProcess(path, os) {
  const rules = IGNORE_RULES[os] || [];
  return rules.some(rule =>
    typeof rule === 'function' ? rule(path) : rule.test(path)
  );
}

// This is what we're going to do:
if (shouldIgnoreProcess(_path, process.platform)) continue;
```

Agree this is much cleaner?

XenHat commented on src/process/index.js:
```diff
@@ -1,18 +1,162 @@
+    /\/bin\/dolphin$/, // KDE, not emulator
+  ],
+  windows: [
+    /^C:\\Windows/i,
```
Thank you! I'll look at this soon, locally if nothing else.

XenHat reviewed (commented):
Looks good so far. I tried to reduce as many changes as possible.

XenHat committed:
process: further improve game detection for Linux
changes:
 - several performance improvements
 - match against more (malformed?) game database entries
 - added simple unit tests

XenHat committed:
process: add Black Mesa for Linux

XenHat committed:
process: add Counter-Strike 2 for Linux

XenHat committed:
Add soulframe executable

This may or may not be reverted upon request from Digital Extremes as
the game is still in development.

XenHat committed:
process: ignore C:\Windows wine processes on Linux

XenHat committed:
process: match steam ID when available
